api/base_controller.go

```diff
@@ -0,0 +1,226 @@
+	return []web.Route{
+		{
+			Endpoint: web.Endpoint{
+				Method: http.MethodPost,
+				Path:   c.resourceBaseURL,
+			},
+			Handler: c.CreateObject,
+		},
+		{
+			Endpoint: web.Endpoint{
+				Method: http.MethodGet,
+				Path:   fmt.Sprintf("%s/{%s}", c.resourceBaseURL, PathParamID),
+			},
+			Handler: c.GetSingleObject,
+		},
+		{
+			Endpoint: web.Endpoint{
+				Method: http.MethodGet,
+				Path:   c.resourceBaseURL,
+			},
+			Handler: c.ListObjects,
+		},
+		{
+			Endpoint: web.Endpoint{
+				Method: http.MethodDelete,
+				Path:   c.resourceBaseURL,
+			},
+			Handler: c.DeleteObjects,
+		},
+		{
+			Endpoint: web.Endpoint{
+				Method: http.MethodDelete,
+				Path:   fmt.Sprintf("%s/{%s}", c.resourceBaseURL, PathParamID),
+			},
+			Handler: c.DeleteSingleObject,
+		},
+		{
+			Endpoint: web.Endpoint{
+				Method: http.MethodPatch,
+				Path:   fmt.Sprintf("%s/{%s}", c.resourceBaseURL, PathParamID),
+			},
+			Handler: c.PatchObject,
+		},
+	}
+}
+
+func (c *BaseController) CreateObject(r *web.Request) (*web.Response, error) {
+	ctx := r.Context()
+	log.C(ctx).Debugf("Creating new %s", c.objectType)
+
+	result := c.objectBlueprint()
+	if err := util.BytesToObject(r.Body, result); err != nil {
+		return nil, err
+	}
+
+	if result.GetID() == "" {
+		UUID, err := uuid.NewV4()
+		if err != nil {
+			return nil, fmt.Errorf("could not generate GUID for %s: %s", c.objectType, err)
+		}
+		result.SetID(UUID.String())
+	}
+	currentTime := time.Now().UTC()
+	result.SetCreatedAt(currentTime)
+	result.SetUpdatedAt(currentTime)
+
+	id, err := c.repository.Create(ctx, result)
+	if err != nil {
+		return nil, err
+	}
+	result.SetID(id)
+	return web.NewJSONResponse(http.StatusCreated, result)
+}
+
+func (c *BaseController) DeleteObjects(r *web.Request) (*web.Response, error) {
+	ctx := r.Context()
+	log.C(ctx).Debugf("Deleting %ss...", c.objectType)
+
+	criteria := query.CriteriaForContext(ctx)
+	if _, err := c.repository.Delete(ctx, c.objectType, criteria...); err != nil {
+		return nil, err
+	}
+
+	return web.NewJSONResponse(http.StatusOK, map[string]string{})
+}
+
+func (c *BaseController) DeleteSingleObject(r *web.Request) (*web.Response, error) {
+	objectID := r.PathParams[PathParamID]
+	ctx := r.Context()
+	log.C(ctx).Debugf("Deleting %s with id %s", c.objectType, objectID)
+
+	byID := query.ByField(query.EqualsOperator, "id", objectID)
+	ctx, err := query.AddCriteria(r.Context(), byID)
+	if err != nil {
+		return nil, err
+	}
+	r.Request = r.WithContext(ctx)
+	return c.DeleteObjects(r)
+}
+
+func (c *BaseController) GetSingleObject(r *web.Request) (*web.Response, error) {
+	objectID := r.PathParams[PathParamID]
+	ctx := r.Context()
+	log.C(ctx).Debugf("Getting %s with id %s", c.objectType, objectID)
+
+	object, err := c.repository.Get(ctx, c.objectType, objectID)
+	if err != nil {
+		return nil, util.HandleStorageError(err, string(c.objectType))
+	}
+	if secured, ok := object.(types.Secured); ok {
```

Comment:
We could save about 10 lines from this file if we extracted this common logic in a function:

```go
func stripCredentials(ctx context.Context, object types.Object) {
	if obj, ok := object.(types.Secured); ok {
		obj.SetCredentials(nil)
	} else {
		log.C(ctx).Debugf("Object of type %s with id %s is not secured, so no credentials are cleaned up on response", object.GetType(), object.GetID())
	}
}
```